README.md
# flutter_persist_http
package design for making your http request easy ,persist your http response and easy to parse respon (json )to model object.
## Getting Started
This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with flutter _persist_http, view our 
[online documentation](https://keshavkeyal.github.io/), which offers tutorials.

## Let see feature of this package 
 🔥Save your http api data  
 🔥Save your time required to parse json 
 🔥Saving list of map obj & single map obj Saving
 🔥Your http Get method with returning objwhich can easily be      parsed to flutter obj.
 🔥Reduce boilerplate code of http 
 🔥Show data when user is offline
 🔥HttpMap class for static method to parse your response          directely to map obj & also thier are method which gives only   response in case u want to apply condition to response and      then u can simply   

CHANGELOG.md
## [0.0.1] - Initial release.

* Save your http request.
* Parse your http request with dart <Map> object.
* Get saved data of http request with dart <Map> object.

## [0.0.2] - Some minor bug fixes.

* Easy to save dart <Map> object.
* Some Minor bug fixes.



test/flutter_persist_http_test.dart
import 'package:flutter_test/flutter_test.dart';
void main() {
  test('adds one to input values', () {
  
  });
}


lib/src/method.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_persist_http/src/exception.dart';
class HttpMap {
  static Future<dynamic> getHttpMap(String url) async {
    String resp;
    try {
      final response = await http.get(url);
      switch (response.statusCode) {
        case 200:
          print(response.statusCode);
          print("getting data" + response.body);
          resp = response.body;
          break;
        case 400:
          print(response.body);
          throw BadRequestException(response.body.toString());
          // print(response.body);
          break;
        case 401:
          print(response.body);
          break;
        case 403:
          print(response.body);
          throw UnauthorisedException(response.body.toString());
          break;
        case 500:
          print(response.body);
          break;
        default:
          print(response.body);
          throw FetchDataException(
              'Error occured while Communication with Server with StatusCode : ${response.statusCode}');
      }
    } catch (e) {
      print('Error ${e.toString()}');
    }
    final parsed =HttpMap.responseToMap(resp);
    return parsed;
  }

  static Future<dynamic> postHttpMap(
      String url, Map<String, dynamic> body) async {
    String resp;
   await http.post(url, body: jsonEncode(body)).then((result) {
      switch (result.statusCode) {
        case 200:
          resp = result.body;
          print(result.statusCode);
          print("getting data" + result.body);
          break;
        // return result.body;
        case 400:
          resp = result.body;
          throw BadRequestException(result.body.toString());
          break;
        case 401:
          break;
        case 403:
          throw UnauthorisedException(result.body.toString());
          break;
        case 500:
          break;
        default:
          resp = result.body;
          throw FetchDataException(
              'Error occured while Communication with Server with StatusCode : ${result.statusCode}');
      }
    }).catchError((onError) {
      print(onError);
    });
   final parsed =HttpMap.responseToMap(resp);
    return parsed;
  }

  static Future<dynamic> httppostResponse(
      String url, Map<String, dynamic> body) async {
    String resp;
    print(body);
   await http.post(url, body: jsonEncode(body)).then((result) {
      switch (result.statusCode) {
        case 200:
          resp = result.body;
          print(result.statusCode);
          print("getting data" + result.body);
          break;
        case 400:
          resp = result.body;
          throw BadRequestException(result.body.toString());
          break;
        case 401:
          break;
        case 403:
          throw UnauthorisedException(result.body.toString());
          break;
        case 500:
          break;
        default:
          resp = result.body;
          throw FetchDataException(
              'Error occured while Communication with Server with StatusCode : ${result.statusCode}');
      }
    }).catchError((onError) {
      print(onError);
    });
    return resp;
  }
  static Future<dynamic> httpGetResponse(String url) async {
    String resp;
   await http.get(url).then((result) {
      switch (result.statusCode) {
        case 200:
          resp = result.body;
          print(result.statusCode);
          print("getting data" + result.body);
          break;
        case 400:
          resp = result.body;
          throw BadRequestException(result.body.toString());
          break;
        case 401:
          break;
        case 403:
          throw UnauthorisedException(result.body.toString());
          break;
        case 500:
          break;
        default:
          resp = result.body;
          throw FetchDataException(
              'Error occured while Communication with Server with StatusCode : ${result.statusCode}');
      }
    }).catchError((onError) {
      print(onError);
    });
    return resp;
  }
  static dynamic responseToMap(String resp)  {
    final parsed = json.decode(resp).cast<Map<String, dynamic>>();
    return parsed;
  }  
}

lib/src/persist.dart
import 'dart:io';
import 'dart:convert';
import 'package:flutter_persist_http/flutter_persist_http.dart';
import 'package:http/http.dart' as http;
class Persist {
  File jsonFile;
  final Directory dir;
  String jsonFileName;
  bool fileExists = false;
  final String apiName;
  Persist(this.dir, this.apiName) {
    this.jsonFileName = apiName + ".json";
    jsonFile = File(dir.path + "/" + jsonFileName);
    fileExists = jsonFile.existsSync();
    if (fileExists) {
      print(jsonDecode(jsonFile.readAsStringSync()));
    } else {
      print("file not created");
    }
  }
  void createFile(String content, Directory dir, String fileName) {
    print("Creating file!");
    File file = File(dir.path + "/" + fileName);
    file.createSync();
    fileExists = true;
    file.writeAsStringSync(content);
  }
  void saveJsonData(String content) {
    print("Writing to file!");
    if (fileExists) {
      print("File exists");
      print("file writing data" + content);
      jsonFile.writeAsStringSync(content);
      print("save data");
    } else {
      print("File does not exist!");
      createFile(content, dir, jsonFileName);
    }
  }
  Future<String> saveHttpGet(String url) async {
    String resp;
    try {
      final response = await http.get(url);
      switch (response.statusCode) {
        case 001:
          resp = response.body;
          print(response.statusCode);
          print("saving your data..." + response.body);
          saveJsonData(response.body);
          break;
        case 200:
          resp = response.body;
          print(response.statusCode);
          print("saving your data..." + response.body);
          saveJsonData(response.body);
          //return response.body;
          break;
        case 400:
          resp = response.body;
          throw BadRequestException(response.body.toString());
          break;
        case 401:
          resp = response.body;
          break;
        case 403:
          resp = response.body;
          throw UnauthorisedException(response.body.toString());
          break;
        case 500:
          resp = response.body;
          break;
        default:
          throw FetchDataException(
              'Error occured while Communication with Server with StatusCode : ${response.statusCode}');
      }
    } catch (e) {
      print('Error ${e.toString()}');
    }
    return resp;
  }
  Future<String> saveHttpPost(String url, Map<String, dynamic> body) async {
    String resp;
    print(body);
    http.post(url, body: jsonEncode(body)).then((result) {
      switch (result.statusCode) {
        case 200:
          resp = result.body;
          print(result.statusCode);
          print("saving your data..." + result.body);
          saveJsonData(result.body);
          break;
        // return result.body;
        case 400:
          resp = result.body;
          throw BadRequestException(result.body.toString());
          break;
        case 401:
          break;
        case 403:
          throw UnauthorisedException(result.body.toString());
          break;
        case 500:
          break;
        default:
          resp = result.body;
          throw FetchDataException(
              'Error occured while Communication with Server with StatusCode : ${result.statusCode}');
      }
    }).catchError((onError) {
      print(onError);
    });
    return resp;
  }
/*  Future<String> getData() async {
    String k;
    if(fileExists)
    k= json.decode(jsonFile.readAsStringSync());
  
  else
  print("file not found");
  return k;
  }*/
  Future<dynamic> getMappedDataFromStorage() async {
    if (fileExists) {
      final parsed =
          json.decode(jsonFile.readAsStringSync()).cast<Map<String, dynamic>>();
      return parsed;
    }
    else{
    print("file not found");
    return "file not found";
    }
  }
  void addNewListMappdedData(List<Map<String, dynamic>> map) {
    print("Writing to file!");
    if (fileExists) {
      List<Map<String, dynamic>> jsonFileContent =
          json.decode(jsonFile.readAsStringSync()).cast<Map<String, dynamic>>();
      jsonFileContent.addAll(map);
      jsonFile.writeAsStringSync(jsonEncode(jsonFileContent));
      print("save data");
    } else {
      print("File does not exist!");
      createFile(jsonEncode(map), dir, jsonFileName);
    }
  }
  void addNewMappdedData(Map<String, dynamic> map) {
    print("Writing to file!");
    if (fileExists) {
      List<Map<String, dynamic>> jsonFileContent =
          json.decode(jsonFile.readAsStringSync());
      jsonFileContent.add(map);
      jsonFile.writeAsStringSync(jsonEncode(jsonFileContent));
      print("save data");
    } else {
      print("File does not exist!");
      createFile(jsonEncode(map), dir, jsonFileName);
    }
  }
  bool isFileExist() {
    return fileExists;
  }
  Future<dynamic> saveHttpGetMap(String url) async {
    String resp;
    try {
      final response = await http.get(url);
      switch (response.statusCode) {
        case 001:
          resp = response.body;
          print(response.statusCode);
          print("saving your data..." + response.body);
          saveJsonData(response.body);
          break;
        case 200:
          resp = response.body;
          print(response.statusCode);
          print("saving your data..." + response.body);
          saveJsonData(response.body);
          //return response.body;
          break;
        case 400:
          resp = response.body;
          throw BadRequestException(response.body.toString());
          break;
        case 401:
          resp = response.body;
          break;
        case 403:
          resp = response.body;
          throw UnauthorisedException(response.body.toString());
          break;
        case 500:
          resp = response.body;
          break;
        default:
          throw FetchDataException(
              'Error occured while Communication with Server with StatusCode : ${response.statusCode}');
      }
    } catch (e) {
      print('Error ${e.toString()}');
    }
    final parsed=HttpMap.responseToMap(resp);
    return parsed;
  }
  Future<dynamic> saveHttpPostMap(
      String url, Map<String, dynamic> body) async {
    String resp;
    print(body);
    http.post(url, body: jsonEncode(body)).then((result) {
      switch (result.statusCode) {
        case 200:
          resp = result.body;
          print(result.statusCode);
          print("saving your data..." + result.body);
          saveJsonData(result.body);
          break;
        // return result.body;
        case 400:
          resp = result.body;
          throw BadRequestException(result.body.toString());
          break;
        case 401:
          break;
        case 403:
          throw UnauthorisedException(result.body.toString());
          break;
        case 500:
          break;
        default:
          resp = result.body;
          throw FetchDataException(
              'Error occured while Communication with Server with StatusCode : ${result.statusCode}');
      }
    }).catchError((onError) {
      print(onError);
    });
        final parsed=HttpMap.responseToMap(resp);
    return parsed;
  }
}